crates/turbo-tasks/src/util.rs

```diff
@@ -47,6 +49,41 @@ impl From<Error> for SharedError {
     }
 }
 
+impl PartialEq for SharedError {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.inner, &other.inner)
+    }
+}
+
+impl Eq for SharedError {}
+
+impl Serialize for SharedError {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let mut v = vec![self.to_string()];
+        let mut source = self.source();
+        while let Some(s) = source {
+            v.push(s.to_string());
+            source = s.source();
+        }
+        Serialize::serialize(&v, serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for SharedError {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        use serde::de::Error;
+        let mut messages = <Vec<String>>::deserialize(deserializer)?;
+        let mut e = match messages.pop() {
+            Some(e) => anyhow!(e),
+            None => return Err(Error::custom("expected at least 1 error message")),
+        };
+        while let Some(message) = messages.pop() {
+            e = e.context(message);
+        }
+        Ok(SharedError::new(e))
+    }
+}
```

Comment:
Not sure if this is a good idea. It's not perfectly restoring error types. Maybe we should return an Serialization error when there is an non-anyhow error in the chain.